File: typescript/packages/core/tests/memory.test.ts
/**
 * Unit tests for the canonical MemoryProvider (spore-core issue #8).
 *
 * Mirrors `rust/crates/spore-core/src/memory.rs#tests` — same rules,
 * same verdicts, parallel structure.
 */

import { describe, expect, it } from "vitest";

import { memory, SessionId } from "../src/index.js";

const { StandardMemoryProvider, MemoryId, Timestamp, MemoryError } = memory;

type SemanticMemory = memory.SemanticMemory;
type EpisodicMemory = memory.EpisodicMemory;
type MemoryQuery = memory.MemoryQuery;

// ── Builders ────────────────────────────────────────────────────────────────

function ts(s: string): memory.Timestamp {
  return new Timestamp(s);
}

function sem(id: string, content: string): SemanticMemory {
  return {
    id: new MemoryId(id),
    content,
    source: { kind: "manual" },
    domain: null,
    version: 1,
    previous_versions: [],
    created_at: ts("2026-05-16T00:00:00Z"),
    updated_at: ts("2026-05-16T00:00:00Z"),
    status: { kind: "active" },
  };
}

function epi(id: string, session: string, content: string): EpisodicMemory {
  return {
    id: new MemoryId(id),
    session_id: new SessionId(session),
    content,
    created_at: ts("2026-05-16T00:00:00Z"),
    tags: [],
  };
}

function makeQuery(overrides: Partial<MemoryQuery> & { task_instruction: string }): MemoryQuery {
  return {
    task_instruction: overrides.task_instruction,
    domain: overrides.domain ?? null,
    session_id: overrides.session_id ?? null,
    min_relevance: overrides.min_relevance ?? 0.5,
    max_items: overrides.max_items ?? 10,
  };
}

describe("MemoryProvider — separate stores", () => {
  it("episodic and semantic use separate stores", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeEpisodic(epi("e1", "s1", "ran tests"));
    await mp.storeSemantic(sem("g1", "always run tests"), "reject");

    const eps = await mp.getEpisodic(new SessionId("s1"));
    expect(eps).toHaveLength(1);
    // Episodic id should not be retrievable as semantic.
    await expect(mp.getSemantic(new MemoryId("e1"))).rejects.toBeInstanceOf(MemoryError);
    // Semantic id should not appear under any session's episodics.
    const none = await mp.getEpisodic(new SessionId("g1"));
    expect(none).toEqual([]);
  });
});

describe("MemoryProvider — semantic replace", () => {
  it("replace archives previous and bumps version", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeSemantic(sem("g1", "v1 content"), "reject");

    const v2 = sem("g1", "v2 content");
    v2.version = 1; // caller version is ignored; provider bumps from existing.
    await mp.storeSemantic(v2, "replace");

    const current = await mp.getSemantic(new MemoryId("g1"));
    expect(current.content).toBe("v2 content");
    expect(current.version).toBe(2);
    expect(current.previous_versions).toHaveLength(1);

    const history = await mp.getVersionHistory(new MemoryId("g1"));
    expect(history).toHaveLength(2);
    expect(history[0]!.content).toBe("v2 content");
    expect(history[1]!.content).toBe("v1 content");
  });

  it("replace chains versions across multiple updates", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeSemantic(sem("g1", "v1"), "reject");
    await mp.storeSemantic(sem("g1", "v2"), "replace");
    await mp.storeSemantic(sem("g1", "v3"), "replace");
    const cur = await mp.getSemantic(new MemoryId("g1"));
    expect(cur.version).toBe(3);
    const history = await mp.getVersionHistory(new MemoryId("g1"));
    expect(history).toHaveLength(3);
  });
});

describe("MemoryProvider — reject", () => {
  it("reject on conflict errors and leaves original untouched", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeSemantic(sem("g1", "first"), "reject");
    await expect(mp.storeSemantic(sem("g1", "second"), "reject")).rejects.toMatchObject({
      kind: "merge_conflict",
    });
    const cur = await mp.getSemantic(new MemoryId("g1"));
    expect(cur.content).toBe("first");
  });
});

describe("MemoryProvider — append", () => {
  it("append concatenates without new version", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeSemantic(sem("g1", "a"), "reject");
    await mp.storeSemantic(sem("g1", "b"), "append");
    const cur = await mp.getSemantic(new MemoryId("g1"));
    expect(cur.content).toBe("ab");
    expect(cur.version).toBe(1);
    expect(cur.previous_versions).toEqual([]);
  });
});

describe("MemoryProvider — validation", () => {
  it("empty content fails validation", async () => {
    const mp = new StandardMemoryProvider();
    await expect(mp.storeSemantic(sem("g1", "   "), "reject")).rejects.toMatchObject({
      kind: "validation_failed",
    });
    await expect(mp.storeEpisodic(epi("e1", "s1", ""))).rejects.toMatchObject({
      kind: "validation_failed",
    });
  });
});

describe("MemoryProvider — meta-agent forced review", () => {
  it("meta-agent memories forced to pending_review", async () => {
    const mp = new StandardMemoryProvider();
    const m = sem("g1", "proposed skill");
    m.source = { kind: "meta_agent_proposed", approved_by: null };
    m.status = { kind: "active" }; // caller dishonestly sets Active
    await mp.storeSemantic(m, "reject");
    const stored = await mp.getSemantic(new MemoryId("g1"));
    expect(stored.status.kind).toBe("pending_review");
  });
});

describe("MemoryProvider — query", () => {
  it("scores, filters by min_relevance, and sorts descending", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeSemantic(sem("g1", "rust async tokio runtime"), "reject");
    await mp.storeSemantic(sem("g2", "python pytest fixtures"), "reject");
    await mp.storeSemantic(sem("g3", "unrelated cooking recipe"), "reject");

    const res = await mp.query(
      makeQuery({ task_instruction: "rust tokio async", min_relevance: 0.1, max_items: 10 }),
    );
    expect(res.length).toBeGreaterThan(0);
    expect(res[0]!.memory.id.asString()).toBe("g1");
    for (let i = 1; i < res.length; i += 1) {
      expect(res[i - 1]!.relevance_score).toBeGreaterThanOrEqual(res[i]!.relevance_score);
    }
  });

  it("excludes deprecated memories", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeSemantic(sem("g1", "rust tokio"), "reject");
    await mp.deprecate(new MemoryId("g1"), "obsolete");
    const res = await mp.query(
      makeQuery({ task_instruction: "rust tokio", min_relevance: 0.0, max_items: 10 }),
    );
    expect(res).toEqual([]);
  });

  it("respects min_relevance and max_items", async () => {
    const mp = new StandardMemoryProvider();
    for (let i = 0; i < 5; i += 1) {
      await mp.storeSemantic(sem(`g${i}`, `alpha beta gamma ${i}`), "reject");
    }
    const none = await mp.query(
      makeQuery({ task_instruction: "alpha beta gamma", min_relevance: 0.99, max_items: 10 }),
    );
    expect(none).toEqual([]);
    const capped = await mp.query(
      makeQuery({ task_instruction: "alpha beta gamma", min_relevance: 0.0, max_items: 2 }),
    );
    expect(capped).toHaveLength(2);
  });

  it("filters by domain", async () => {
    const mp = new StandardMemoryProvider();
    const a = sem("a", "shared content");
    a.domain = "rust";
    const b = sem("b", "shared content");
    b.domain = "python";
    await mp.storeSemantic(a, "reject");
    await mp.storeSemantic(b, "reject");
    const res = await mp.query(
      makeQuery({
        task_instruction: "shared content",
        domain: "rust",
        min_relevance: 0.0,
        max_items: 10,
      }),
    );
    expect(res).toHaveLength(1);
    expect(res[0]!.memory.id.asString()).toBe("a");
  });
});

describe("MemoryProvider — lifecycle", () => {
  it("deprecate sets status", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeSemantic(sem("g1", "x"), "reject");
    await mp.deprecate(new MemoryId("g1"), "no longer needed");
    const m = await mp.getSemantic(new MemoryId("g1"));
    expect(m.status.kind).toBe("deprecated");
    if (m.status.kind === "deprecated") {
      expect(m.status.reason).toBe("no longer needed");
    }
  });

  it("deprecate unknown id returns not_found", async () => {
    const mp = new StandardMemoryProvider();
    await expect(mp.deprecate(new MemoryId("nope"), "r")).rejects.toMatchObject({
      kind: "not_found",
    });
  });

  it("markPendingReview changes status", async () => {
    const mp = new StandardMemoryProvider();
    await mp.storeSemantic(sem("g1", "x"), "reject");
    await mp.markPendingReview(new MemoryId("g1"));
    const m = await mp.getSemantic(new MemoryId("g1"));
    expect(m.status.kind).toBe("pending_review");
  });
});

describe("MemoryProvider — NotFound / empty", () => {
  it("getSemantic unknown returns not_found", async () => {
    const mp = new StandardMemoryProvider();
    await expect(mp.getSemantic(new MemoryId("nope"))).rejects.toMatchObject({
      kind: "not_found",
    });
  });

  it("getEpisodic unknown session returns empty", async () => {
    const mp = new StandardMemoryProvider();
    const res = await mp.getEpisodic(new SessionId("none"));
    expect(res).toEqual([]);
  });
});

describe("MemoryProvider — episodic ordering", () => {
  it("preserves insertion order across multiple writes", async () => {
    const mp = new StandardMemoryProvider();
    for (let i = 0; i < 5; i += 1) {
      await mp.storeEpisodic(epi(`e${i}`, "s1", `event ${i}`));
    }
    const eps = await mp.getEpisodic(new SessionId("s1"));
    expect(eps).toHaveLength(5);
    eps.forEach((e, i) => {
      expect(e.id.asString()).toBe(`e${i}`);
    });
  });
});

File: go/spore-core/memory/memory_test.go
package memory

import (
	"context"
	"errors"
	"fmt"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

func ts(s string) Timestamp { return Timestamp(s) }

func sem(id, content string) SemanticMemory {
	return SemanticMemory{
		ID:               MemoryID(id),
		Content:          content,
		Source:           NewSourceManual(),
		Domain:           nil,
		Version:          1,
		PreviousVersions: nil,
		CreatedAt:        ts("2026-05-16T00:00:00Z"),
		UpdatedAt:        ts("2026-05-16T00:00:00Z"),
		Status:           NewStatusActive(),
	}
}

func epi(id, session, content string) EpisodicMemory {
	return EpisodicMemory{
		ID:        MemoryID(id),
		SessionID: sporecore.SessionID(session),
		Content:   content,
		CreatedAt: ts("2026-05-16T00:00:00Z"),
		Tags:      nil,
	}
}

func isKind(err error, kind MemoryErrorKind) bool {
	var me *MemoryError
	if !errors.As(err, &me) {
		return false
	}
	return me.Kind == kind
}

// ── Rule: Episodic and semantic stored/retrieved separately ─────────────────

func TestEpisodicAndSemanticUseSeparateStores(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	if _, err := mp.StoreEpisodic(ctx, epi("e1", "s1", "ran tests")); err != nil {
		t.Fatal(err)
	}
	if _, err := mp.StoreSemantic(ctx, sem("g1", "always run tests"), MergeStrategyReject); err != nil {
		t.Fatal(err)
	}

	eps, err := mp.GetEpisodic(ctx, sporecore.SessionID("s1"))
	if err != nil {
		t.Fatal(err)
	}
	if len(eps) != 1 {
		t.Fatalf("want 1 episodic, got %d", len(eps))
	}
	if _, err := mp.GetSemantic(ctx, MemoryID("e1")); err == nil {
		t.Fatalf("expected NotFound for episodic id in semantic store")
	}
	other, _ := mp.GetEpisodic(ctx, sporecore.SessionID("g1"))
	if len(other) != 0 {
		t.Fatalf("semantic id leaked into episodic")
	}
}

// ── Rule: Replace creates a new version, retains previous ───────────────────

func TestReplaceArchivesPreviousAndBumpsVersion(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	if _, err := mp.StoreSemantic(ctx, sem("g1", "v1 content"), MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	v2 := sem("g1", "v2 content")
	v2.Version = 1
	if _, err := mp.StoreSemantic(ctx, v2, MergeStrategyReplace); err != nil {
		t.Fatal(err)
	}

	cur, err := mp.GetSemantic(ctx, MemoryID("g1"))
	if err != nil {
		t.Fatal(err)
	}
	if cur.Content != "v2 content" {
		t.Fatalf("content: %q", cur.Content)
	}
	if cur.Version != 2 {
		t.Fatalf("version: %d", cur.Version)
	}
	if len(cur.PreviousVersions) != 1 {
		t.Fatalf("previous_versions: %d", len(cur.PreviousVersions))
	}

	hist, err := mp.GetVersionHistory(ctx, MemoryID("g1"))
	if err != nil {
		t.Fatal(err)
	}
	if len(hist) != 2 || hist[0].Content != "v2 content" || hist[1].Content != "v1 content" {
		t.Fatalf("unexpected history: %+v", hist)
	}
}

func TestReplaceChainsVersionsAcrossMultipleUpdates(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	if _, err := mp.StoreSemantic(ctx, sem("g1", "v1"), MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	if _, err := mp.StoreSemantic(ctx, sem("g1", "v2"), MergeStrategyReplace); err != nil {
		t.Fatal(err)
	}
	if _, err := mp.StoreSemantic(ctx, sem("g1", "v3"), MergeStrategyReplace); err != nil {
		t.Fatal(err)
	}
	cur, _ := mp.GetSemantic(ctx, MemoryID("g1"))
	if cur.Version != 3 {
		t.Fatalf("version: %d", cur.Version)
	}
	hist, _ := mp.GetVersionHistory(ctx, MemoryID("g1"))
	if len(hist) != 3 {
		t.Fatalf("history len: %d", len(hist))
	}
}

// ── Rule: Reject returns MergeConflict ──────────────────────────────────────

func TestRejectOnConflictErrors(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	if _, err := mp.StoreSemantic(ctx, sem("g1", "first"), MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	_, err := mp.StoreSemantic(ctx, sem("g1", "second"), MergeStrategyReject)
	if !isKind(err, ErrKindMergeConflict) {
		t.Fatalf("want MergeConflict, got %v", err)
	}
	cur, _ := mp.GetSemantic(ctx, MemoryID("g1"))
	if cur.Content != "first" {
		t.Fatalf("original mutated: %q", cur.Content)
	}
}

// ── Rule: Append concatenates in place, no new version ─────────────────────

func TestAppendConcatenatesWithoutNewVersion(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	if _, err := mp.StoreSemantic(ctx, sem("g1", "a"), MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	if _, err := mp.StoreSemantic(ctx, sem("g1", "b"), MergeStrategyAppend); err != nil {
		t.Fatal(err)
	}
	cur, _ := mp.GetSemantic(ctx, MemoryID("g1"))
	if cur.Content != "ab" {
		t.Fatalf("content: %q", cur.Content)
	}
	if cur.Version != 1 {
		t.Fatalf("version: %d", cur.Version)
	}
	if len(cur.PreviousVersions) != 0 {
		t.Fatalf("previous_versions: %d", len(cur.PreviousVersions))
	}
}

// ── Rule: Writes validated ─────────────────────────────────────────────────

func TestEmptyContentFailsValidation(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	_, err := mp.StoreSemantic(ctx, sem("g1", "   "), MergeStrategyReject)
	if !isKind(err, ErrKindValidationFailed) {
		t.Fatalf("want ValidationFailed, got %v", err)
	}
	_, err = mp.StoreEpisodic(ctx, epi("e1", "s1", ""))
	if !isKind(err, ErrKindValidationFailed) {
		t.Fatalf("want ValidationFailed for episodic, got %v", err)
	}
}

// ── Rule: MetaAgentProposed forced to PendingReview ────────────────────────

func TestMetaAgentMemoriesForcedToPendingReview(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	m := sem("g1", "proposed skill")
	m.Source = NewSourceMetaAgentProposed(nil)
	m.Status = NewStatusActive() // caller dishonestly sets Active
	if _, err := mp.StoreSemantic(ctx, m, MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	stored, _ := mp.GetSemantic(ctx, MemoryID("g1"))
	if stored.Status.Kind != StatusKindPendingReview {
		t.Fatalf("want PendingReview, got %v", stored.Status.Kind)
	}
}

// ── Rule: query scores, filters, sorts ─────────────────────────────────────

func TestQueryScoresFiltersAndSorts(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	for _, m := range []SemanticMemory{
		sem("g1", "rust async tokio runtime"),
		sem("g2", "python pytest fixtures"),
		sem("g3", "unrelated cooking recipe"),
	} {
		if _, err := mp.StoreSemantic(ctx, m, MergeStrategyReject); err != nil {
			t.Fatal(err)
		}
	}
	q := MemoryQuery{TaskInstruction: "rust tokio async", MinRelevance: 0.1, MaxItems: 10}
	res, err := mp.Query(ctx, q)
	if err != nil {
		t.Fatal(err)
	}
	if len(res) == 0 {
		t.Fatal("expected results")
	}
	if res[0].Memory.ID != MemoryID("g1") {
		t.Fatalf("top result: %v", res[0].Memory.ID)
	}
	for i := 1; i < len(res); i++ {
		if res[i-1].RelevanceScore < res[i].RelevanceScore {
			t.Fatalf("not sorted desc")
		}
	}
}

func TestQueryExcludesDeprecatedMemories(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	if _, err := mp.StoreSemantic(ctx, sem("g1", "rust tokio"), MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	if err := mp.Deprecate(ctx, MemoryID("g1"), "obsolete"); err != nil {
		t.Fatal(err)
	}
	res, _ := mp.Query(ctx, MemoryQuery{TaskInstruction: "rust tokio", MinRelevance: 0, MaxItems: 10})
	if len(res) != 0 {
		t.Fatalf("deprecated returned: %+v", res)
	}
}

func TestQueryRespectsMinRelevanceAndMaxItems(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	for i := 0; i < 5; i++ {
		_, err := mp.StoreSemantic(ctx, sem(fmt.Sprintf("g%d", i), fmt.Sprintf("alpha beta gamma %d", i)), MergeStrategyReject)
		if err != nil {
			t.Fatal(err)
		}
	}
	res, _ := mp.Query(ctx, MemoryQuery{TaskInstruction: "alpha beta gamma", MinRelevance: 0.99, MaxItems: 10})
	if len(res) != 0 {
		t.Fatalf("min_relevance not enforced: %d", len(res))
	}
	res, _ = mp.Query(ctx, MemoryQuery{TaskInstruction: "alpha beta gamma", MinRelevance: 0, MaxItems: 2})
	if len(res) != 2 {
		t.Fatalf("max_items not enforced: %d", len(res))
	}
}

func TestQueryFiltersByDomain(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	rust, py := "rust", "python"
	a := sem("a", "shared content")
	a.Domain = &rust
	b := sem("b", "shared content")
	b.Domain = &py
	if _, err := mp.StoreSemantic(ctx, a, MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	if _, err := mp.StoreSemantic(ctx, b, MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	res, _ := mp.Query(ctx, MemoryQuery{
		TaskInstruction: "shared content",
		Domain:          &rust,
		MinRelevance:    0,
		MaxItems:        10,
	})
	if len(res) != 1 || res[0].Memory.ID != MemoryID("a") {
		t.Fatalf("domain filter failed: %+v", res)
	}
}

// ── Rule: Lifecycle — deprecate ────────────────────────────────────────────

func TestDeprecateSetsStatus(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	if _, err := mp.StoreSemantic(ctx, sem("g1", "x"), MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	if err := mp.Deprecate(ctx, MemoryID("g1"), "no longer needed"); err != nil {
		t.Fatal(err)
	}
	m, _ := mp.GetSemantic(ctx, MemoryID("g1"))
	if m.Status.Kind != StatusKindDeprecated || m.Status.Reason != "no longer needed" {
		t.Fatalf("status: %+v", m.Status)
	}
}

func TestDeprecateUnknownIDNotFound(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	err := mp.Deprecate(ctx, MemoryID("nope"), "r")
	if !isKind(err, ErrKindNotFound) {
		t.Fatalf("want NotFound, got %v", err)
	}
}

// ── Rule: mark_pending_review ──────────────────────────────────────────────

func TestMarkPendingReviewChangesStatus(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	if _, err := mp.StoreSemantic(ctx, sem("g1", "x"), MergeStrategyReject); err != nil {
		t.Fatal(err)
	}
	if err := mp.MarkPendingReview(ctx, MemoryID("g1")); err != nil {
		t.Fatal(err)
	}
	m, _ := mp.GetSemantic(ctx, MemoryID("g1"))
	if m.Status.Kind != StatusKindPendingReview {
		t.Fatalf("status: %+v", m.Status)
	}
}

// ── Rule: NotFound errors ──────────────────────────────────────────────────

func TestGetSemanticUnknownReturnsNotFound(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	_, err := mp.GetSemantic(ctx, MemoryID("nope"))
	if !isKind(err, ErrKindNotFound) {
		t.Fatalf("want NotFound, got %v", err)
	}
}

func TestGetEpisodicUnknownSessionReturnsEmpty(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	res, err := mp.GetEpisodic(ctx, sporecore.SessionID("none"))
	if err != nil {
		t.Fatal(err)
	}
	if len(res) != 0 {
		t.Fatalf("expected empty, got %d", len(res))
	}
}

// ── Episodic preserves insertion order ─────────────────────────────────────

func TestEpisodicPreservesOrder(t *testing.T) {
	ctx := context.Background()
	mp := NewStandardMemoryProvider()
	for i := 0; i < 5; i++ {
		if _, err := mp.StoreEpisodic(ctx, epi(fmt.Sprintf("e%d", i), "s1", fmt.Sprintf("event %d", i))); err != nil {
			t.Fatal(err)
		}
	}
	eps, _ := mp.GetEpisodic(ctx, sporecore.SessionID("s1"))
	if len(eps) != 5 {
		t.Fatalf("len: %d", len(eps))
	}
	for i, e := range eps {
		if e.ID != MemoryID(fmt.Sprintf("e%d", i)) {
			t.Fatalf("order broken at %d: %v", i, e.ID)
		}
	}
}

// ── NewMemoryQuery defaults ────────────────────────────────────────────────

func TestNewMemoryQueryDefaults(t *testing.T) {
	q := NewMemoryQuery("hello")
	if q.TaskInstruction != "hello" {
		t.Fatalf("instr: %q", q.TaskInstruction)
	}
	if q.MinRelevance != 0.5 {
		t.Fatalf("min: %v", q.MinRelevance)
	}
	if q.MaxItems != 10 {
		t.Fatalf("max: %v", q.MaxItems)
	}
}
